airbyte-webapp/src/components/ui/DatePicker/DatePicker.tsx

```diff
@@ -0,0 +1,155 @@
+ * 2022-01-01T09:00:00+01:00  - what we give react-datepicker instead, to trick it (User sees 9:00am - good!)
+ */
+export const toEquivalentLocalTime = (input: string): Date | undefined => {
+  if (!input) {
+    return undefined;
+  }
+
+  const date = dayjs.utc(input);
+
+  if (!date?.isValid()) {
+    return undefined;
+  }
+
+  // Get the user's UTC offset based on the local time
+  const browserUtcOffset = dayjs().utcOffset();
+
+  // Convert the selected date into a string which we can use to initialize a new date object.
+  // The second parameter to utcOffset() keeps the same local time, only changing the timezone.
+  const dateInUtcAsString = date.utcOffset(browserUtcOffset, true).format();
+
+  const equivalentDate = dayjs(dateInUtcAsString);
+
+  // dayjs does not 0-pad years when formatting, so it's possible to have an invalid date here
+  // https://github.com/iamkun/dayjs/issues/1745
+  if (!equivalentDate.isValid()) {
+    return undefined;
+  }
+
+  return equivalentDate.toDate();
+};
+
+export interface DatePickerProps {
+  error?: boolean;
+  value: string;
+  onChange: (value: string) => void;
+  withTime?: boolean;
+  disabled?: boolean;
+  onBlur?: (ev: React.FocusEvent<HTMLInputElement>) => void;
+  placeholder?: string;
+}
+
+interface DatePickerButtonTriggerProps {
+  onClick?: () => void;
+}
+
+const DatepickerButton = React.forwardRef<HTMLButtonElement, DatePickerButtonTriggerProps>(({ onClick }, ref) => {
+  const { formatMessage } = useIntl();
+
+  return (
+    <Button
+      className={styles.datepickerButton}
+      aria-label={formatMessage({ id: "form.openDatepicker" })}
+      onClick={onClick}
+      ref={ref}
+      type="button"
+      variant="clear"
+      icon={<FontAwesomeIcon icon={faCalendarAlt} className={styles.dropdownButton} fixedWidth />}
+    />
+  );
+});
+
+export const DatePicker: React.FC<DatePickerProps> = ({
+  disabled,
+  error,
+  onChange,
+  onBlur,
+  placeholder,
+  value = "",
+  withTime = false,
+}) => {
+  const { locale } = useIntl();
+  const datepickerRef = useRef<ReactDatePicker>(null);
+
+  // Additional locales can be registered here as necessary
+  useEffect(() => {
+    switch (locale) {
+      case "en":
+        registerLocale(locale, en);
+        break;
+    }
+  }, [locale]);
+
+  const handleDatepickerChange = useCallback(
+    (val: Date | null) => {
+      const date = dayjs(val);
+      if (!date.isValid()) {
+        onChange("");
+        return;
+      }
+
+      const formattedDate = withTime ? date.utcOffset(0, true).format() : date.format("YYYY-MM-DD");
+      onChange(formattedDate);
+    },
+    [onChange, withTime]
+  );
+
+  const handleInputChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      onChange(e.target.value);
+    },
+    [onChange]
+  );
+
+  const localDate = useMemo(() => toEquivalentLocalTime(value), [value]);
+
+  return (
+    <div className={styles.wrapper}>
+      <Input
```

Comment:
When testing locally, I noticed that if the datepicker is opened by focusing the input, and then a date is clicked in order to fill it in, the UI quickly flashes the `Required` error on the field before going away once it is filled. Here is a screen recording of this:

https://user-images.githubusercontent.com/22731524/204665634-cffbce1d-140f-4ba1-9980-59500dcad7de.mov

This doesn't seem to happen when just clicking on the Datepicker button directly. So my guess is that clicking into the input to focus it, then clicking somewhere in the datepicker popup causes the field to be marked as "touched", and since it doesn't have any value at that point, we show the empty Formik value error (you can reproduce the same situation if you just click on another required input then click away). Then the datepicker actually fills in the field, making it no longer have an error, so it goes away.

Maybe can somehow keep the Input as "focused" or prevent it from being "blurred" to prevent this? Or somehow render the datepicker pop "inside" the Input so that it is not being clicked out of?

Reply:
I think we should solve that by making sure that we only trigger the external `onBlur` method if actually the focus left the wrapping element of the date picker. I.e. we should treat focus within that wrapping element as still focusing this component.

Reply:
Thanks for catching that @lmossman! I took @timroes advice and attached a blur listener to the wrapper div:

```ts
  const handleWrapperBlur = (event: React.FocusEvent<HTMLDivElement>) => {
    if (onBlur && !wrapperRef.current?.matches(":focus-within")) {
      onBlur(event);
    }
  };
```